elastic/fleet-server
# feat: support static tokens to be used for agent enrollment

## What is the problem this PR solves?
This PR adds support for static tokens for an agent enrollment.
The reason we need this capability is to support a flow for serverless security projects when an agent should be installed at the same time as the other project services (kibana, es, fleet). 
To be able to install an agent at that point, we need a predictable token that both the agent and the fleet accept.

## How does this PR solve the problem?
By accepting static tokens, well just the `key` part of the token, fleet server will authenticate a request that have this `key` in the enrollment token it uses and will send back the API key to communicate with ES. For example, passing to the agent token like `base64(0123456789:abcdefg)` the fleet server will accept it if the configuration includes the following
```yaml
inputs:
  - type: fleet-server
    policy.id: "${FLEET_SERVER_POLICY_ID:fleet-server-policy}"
    server:
      auth: statis
      static_policy_tokens:
        enabled: true
        policy_tokens:
          # token.key: policy_id
          abcdefg: 901b70f0-fefa-11ed-aa5e-5974b0535e80
```

Today flow looks something like that 
![image](https://github.com/elastic/fleet-server/assets/23185631/d479710d-1ba5-4011-8e84-f4c08d997154)


And this PR will support the following flow
![image](https://github.com/elastic/fleet-server/assets/23185631/0d15bc26-bed8-4fa1-af84-d86ddc99907d)



## How to test this PR locally
1. Create local stack `elastic-package stack up -s "elasticsearch,kibana,package-registry" --version 8.8.0 -v -d`
2. Build and run fleet server local image ( attached to the network of the stack )
3. Run Elastic-Agent docker image (attached to the same network )
4. See that the Elastic-Agent is healthy and the spawned Cloudbeat sending findings.

Steps 2,3 looks simple but they require more configuration and changes (especially if I want to debug the fleet process)
First building docker image for the server will not work on arm architecture as the `make build-docker` [compiles](https://github.com/elastic/fleet-server/blob/main/Dockerfile#L15) it for amd ( I guess that I was missing something).
Later to command to run the image looks like 
```bash
docker run -it --rm \
    --network elastic-package-stack_default \
    -v $(PWD)/fleet-server.yml:/etc/fleet-server.yml \
    -v $(HOME)/.elastic-package/profiles/default/certs/fleet-server:/etc/ssl/certs \
    -v $(HOME)/.elastic-package/profiles/default/certs/ca-cert.pem:/etc/ssl/certs/elastic-package.pem \
    -e FLEET_TOKEN_POLICY_NAME="Fleet Server (elastic-package)" \
    -e FLEET_URL=https://fleet-server:8220 \
    -e KIBANA_HOST=https://kibana:5601 \
    -e ELASTIC_CONTAINER=true \
    -e FLEET_SERVER_SERVICE_TOKEN=**REDACTED**\
    -e FLEET_SERVER_ENABLE=1 \
    -e ELASTICSEARCH_HOSTS=https://elasticsearch:9200 \
    -e KIBANA_FLEET_SERVICE_TOKEN=**REDACTED** \
    -e KIBANA_FLEET_HOST=https://kibana:5601 \
    -e FLEET_SERVER_ELASTICSEARCH_HOST=https://elasticsearch:9200 \
    -e FLEET_SERVER_HOST=0.0.0.0 \
    -e FLEET_SERVER_CERT_KEY=/etc/ssl/certs/key.pem \
    -e FLEET_SERVER_CERT=/etc/ssl/certs/cert.pem \
    -e KIBANA_FLEET_SETUP=1 \
    -p 8220:8220 \
    --name fleet-server \
    docker.elastic.co/fleet-server/fleet-server:8.9.0
```
Where the change in `fleet-server.yml` was to update `inputs[0].server` with 
```yaml
      static_policy_tokens:
        enabled: true
        policy_tokens:
          # token.key: policy_id
          abcdefg: 901b70f0-fefa-11ed-aa5e-5974b0535e80
```

The command to run Elastic Agent looks like
```bash
docker run -it \
    --network elastic-package-stack_default \
    -v $(HOME)/.elastic-package/profiles/default/certs/ca-cert.pem:/etc/ssl/certs/elastic-package.pem \
    -e FLEET_ENROLLMENT_TOKEN=MDEyMzQ6YWJjZGVmZw== \
    -e FLEET_ENROLL=1 \
    -e FLEET_URL=https://fleet-server:8220 \
    -e KIBANA_HOST=https://kibana:5601 \
    -e ELASTIC_CONTAINER=true \
    docker.elastic.co/elastic-agent/elastic-agent-complete:8.8.0
```

## Design Checklist




## Checklist





## Related issues
- https://github.com/elastic/security-team/issues/6762
- https://github.com/elastic/security-team/issues/6648
